### PR TITLE
Introduce retry on resource conflict

### LINKED ISSUE #3239: Retry on resource conflict
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
This issue is difficult to reproduce but periodically shows up in the logs. It's not clear whether it has any impact on KServe's functionality but it definitely results in logspam and confusion for our users.
```bash
{"level":"error","ts":"2023-11-10T07:55:36Z","msg":"Reconciler error","controller":"inferenceservice","controllerGroup":"serving.kserve.io","controllerKind":"InferenceService","InferenceService":{"name":"sklearn-iris","namespace":"default"},"namespace":"default","name":"sklearn-iris","reconcileID":"f3773c51-ae77-4b94-8f12-66f0a417f7bc","error":"Operation cannot be fulfilled on inferenceservices.serving.kserve.io \"sklearn-iris\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:226"}

```
```bash
{"level":"error","ts":"2023-11-07T10:36:13Z","logger":"KsvcReconciler","msg":"Failed to perform dry-run create of knative service","service":"sklearn-iris-transformer","error":"Operation cannot be fulfilled on services.serving.knative.dev \"sklearn-iris-transformer\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/reconcilers/knative.(*KsvcReconciler).Reconcile\n\tgithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:233\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/components.(*Transformer).Reconcile\n\tgithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/components/transformer.go:200\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice.(*InferenceServiceReconciler).Reconcile\n\tgithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/controller.go:182\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:118\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:314\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:265\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.15.2/pkg/internal/controller/controller.go:226"}

```
Also see issue #3033 
For knative service reconcilation retry logic is already introduced by PR #3035 

**What did you expect to happen:**
No such logs should show up in KServe. We should reattempt the update on resource conflict.

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KServe Version: master
- Kubeflow version:
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube/Kind version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:
Introduces retry logic on resource conflict when updating resources isvc and inferencegraph

**Which issue(s) this PR fixes**:
Fixes #3239

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go [hunk 1]
```diff
@@ -224,25 +224,29 @@ func (r *KsvcReconciler) Reconcile() (*knservingv1.ServiceStatus, error) {
...
+		// Add immutable annotations to avoid validation error during dry-run update.
+		desired.Annotations[knserving.CreatorAnnotation] = existing.Annotations[knserving.CreatorAnnotation]
+		desired.Annotations[knserving.UpdaterAnnotation] = existing.Annotations[knserving.UpdaterAnnotation]
+
+		// Do a dry-run update to avoid diffs generated by default values introduced by knative's defaulter webhook.
+		// This will populate our local knative service object with any default values
+		// that are present on the remote version.
+		if err = r.client.Update(context.TODO(), desired, client.DryRunAll); err != nil {
+			// log only if it is not resource conflict error to avoid spamming
+			if !apierr.IsConflict(err) {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:236]: Is the main change here to wrap setting the resource version in the retry block ?

Comment (yuzisun) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/knative/ksvc_reconciler.go:236]: Also do we still need the outer get ksvc call if it is wrapped in the retry block

**Comment by sivanantha321** [comment]
@yuzisun Hey Dan, It turns out the ksvc status is updated by the knative controller which leads to this race condition.
Here is the audit log  
```
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"RequestResponse","auditID":"fba3850e-7454-412f-b3d2-a57d69379f19","stage":"RequestReceived","requestURI":"/apis/serving.knative.dev/v1/namespaces/default/services/sklearn-iris-predictor/status","verb":"update","user":{"username":"system:serviceaccount:knative-serving:controller","uid":"26966c78-26df-4905-97f5-03ccbf195b7b","groups":["system:serviceaccounts","system:serviceaccounts:knative-serving","system:authenticated"],"extra":{"authentication.kubernetes.io/pod-name":["controller-5649857ccc-5vdfd"],"authentication.kubernetes.io/pod-uid":["51020d7c-4f0b-45d1-8f82-6ae155259211"]}},"sourceIPs":["10.244.0.9"],"userAgent":"controller/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"services","namespace":"default","name":"sklearn-iris-predictor","apiGroup":"serving.knative.dev","apiVersion":"v1","subresource":"status"},"requestReceivedTimestamp":"2023-12-07T07:37:58.868633Z","stageTimestamp":"2023-12-07T07:37:58.868633Z"}

```

File: pkg/controller/v1alpha1/inferencegraph/knative_reconciler.go [hunk 2]
```diff
@@ -90,41 +90,41 @@ func reconcileKsvc(desired *knservingv1.Service, existing *knservingv1.Service)
...
-	// Do a dry-run update to avoid diffs generated because knative defaulter webhook is
-	// called when creating or updating the knative service. This will populate our local knative service object with any default values
-	// that are present on the remote version.
-	if err = r.client.Update(context.TODO(), desired, client.DryRunAll); err != nil {
-		log.Error(err, "Failed to perform dry-run create of knative service", "service", desired.Name)
-		return nil, err
-	}
 
-	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/knative_reconciler.go:94]: it is better to wrap `retry.RetryOnConflict` on `Reconcile` instead if the entire function needs to be in the retry block.

File: pkg/controller/v1beta1/inferenceservice/controller.go [hunk 3]
```diff
@@ -95,193 +94,215 @@ type InferenceServiceReconciler struct {
...
 
-	// Fetch the InferenceService instance
 	isvc := &v1beta1api.InferenceService{}
-	if err := r.Get(ctx, req.NamespacedName, isvc); err != nil {
-		if apierr.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
-			return reconcile.Result{}, nil
+	var result reconcile.Result
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/controller.go:99]: it is better to wrap `retry.RetryOnConflict` on `Reconcile` instead if the entire function needs to be in the retry block.

Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/controller.go]: I do not think we need retry for isvc update as kserve controller is the only one updating it

File: pkg/controller/v1alpha1/inferencegraph/controller.go [hunk 4]
```diff
@@ -116,134 +116,146 @@ func getRouterConfigs(configMap *v1.ConfigMap) (*RouterConfig, error) {
...
 
-	// Fetch the InferenceService instance
 	graph := &v1alpha1api.InferenceGraph{}
-	if err := r.Get(ctx, req.NamespacedName, graph); err != nil {
-		if apierr.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
-			return reconcile.Result{}, nil
+	var result ctrl.Result
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
Comment (yuzisun) [inline pkg/controller/v1alpha1/inferencegraph/controller.go]: We shouldn't have conflicts for inference graph reconciliation.

**Comment by yuzisun** [comment]
@sivanantha321 The scope of the change is a bit large, let's focus on fixing Knative service reconciling first and leave the top level `InferenceService` and `InferenceGraph` for now see if it is really needed.

**Pull request merged**